src/pytest_split/algorithms.py

```diff
@@ -0,0 +1,106 @@
+        List of groups
+    """
+    durations = remove_irrelvant_durations(items, durations)
+    avg_duration_per_test = get_avg_duration_per_test(durations)
+
+    selected: "List[List[nodes.Item]]" = [[] for i in range(splits)]
+    deselected: "List[List[nodes.Item]]" = [[] for i in range(splits)]
+    duration: "List[float]" = [0 for i in range(splits)]
+
+    # create a heap of the form (summed_durations, group_index)
+    heap: "List[Tuple[float, int]]" = [(0, i) for i in range(splits)]
+    heapq.heapify(heap)
+    for item in items:
+        item_duration = durations.get(item.nodeid, avg_duration_per_test)
+
+        # get group with smallest sum
+        summed_durations, group_idx = heapq.heappop(heap)
+        new_group_durations = summed_durations + item_duration
+
+        # store assignment
+        selected[group_idx].append(item)
+        duration[group_idx] = new_group_durations
+        for i in range(splits):
+            if i != group_idx:
+                deselected[i].append(item)
+
+        # store new duration - in case of ties it sorts by the group_idx
+        heapq.heappush(heap, (new_group_durations, group_idx))
+
+    return [test_group(selected=selected[i], deselected=deselected[i], duration=duration[i]) for i in range(splits)]
+
+
+def duration_based_chunks(splits: int, items: "List[nodes.Item]", durations: "Dict[str, float]") -> "List[test_group]":
+    """
+    Split tests into groups by runtime.
+    Ensures tests are split into non-overlapping groups.
+    The original list of test items is split into groups by finding boundary indices i_0, i_1, i_2
+    and creating group_1 = items[0:i_0], group_2 = items[i_0, i_1], group_3 = items[i_1, i_2], ...
+
+    :param splits: How many groups we're splitting in.
+    :param items: Test items passed down by Pytest.
+    :param durations: Our cached test runtimes. Assumes contains timings only of relevant tests
+    :return: List of test_group
+    """
+    durations = remove_irrelvant_durations(items, durations)
+    avg_duration_per_test = get_avg_duration_per_test(durations)
+
+    tests_and_durations = {item: durations.get(item.nodeid, avg_duration_per_test) for item in items}
+    time_per_group = sum(tests_and_durations.values()) / splits
+
+    selected: "List[List[nodes.Item]]" = [[] for i in range(splits)]
+    deselected: "List[List[nodes.Item]]" = [[] for i in range(splits)]
+    duration: "List[float]" = [0 for i in range(splits)]
+
+    group_idx = 0
+    for item in items:
+        if duration[group_idx] >= time_per_group:
+            group_idx += 1
+
+        selected[group_idx].append(item)
+        for i in range(splits):
+            if i != group_idx:
+                deselected[i].append(item)
+        duration[group_idx] += tests_and_durations.pop(item)
+
+    return [test_group(selected=selected[i], deselected=deselected[i], duration=duration[i]) for i in range(splits)]
+
+
+def get_avg_duration_per_test(durations: "Dict[str, float]") -> float:
```

Comment:
This and `remove_irrelvant_durations` should be probably private, could you indicate that with `_` prefix